mlir/lib/Dialect/MIOpen/Generator/Conv2dGenerator.cpp

```diff
@@ -10,63 +10,146 @@
-    } else if (outputDim == 'w') {
-      outputDimension.push_back(outputWidth);
+    // Determine tensor dimensions.
+    return parseConvDims(
+        strToLong("batchsize"), strToLong("in_channels"), strToLong("in_h"),
+        strToLong("in_w"), strToLong("out_channels"), strToLong("out_h"),
+        strToLong("out_w"), strToLong("fil_w"), strToLong("fil_h"));
+  }
+
+  return failure();
+}
+
+LogicalResult Conv2dGenerator::parseConvDims(
+    int64_t batchSize, int64_t inputChannel, int64_t inputHeight,
+    int64_t inputWidth, int64_t outputChannel, int64_t outputHeight,
+    int64_t outputWidth, int64_t filterHeight, int64_t filterWidth) {
+
+  static const std::string filterKeys = "kcyx";
+  int64_t filterVals[] = {outputChannel, inputChannel, filterHeight,
+                          filterWidth};
+
+  static const std::string inputKeys = "nchw";
+  int64_t inputVals[] = {batchSize, inputChannel, inputHeight, inputWidth};
+
+  static const std::string outputKeys = "nkhw";
+  int64_t outputVals[] = {batchSize, outputChannel, outputHeight, outputWidth};
+
+  auto convertLayout = [](char key, const std::string &kmap, int64_t vals[],
+                          auto &dims) {
+    auto keyl = std::tolower(key);
+    auto ii = kmap.find(keyl);
+    if (ii == std::string::npos) {
+      static std::string nchw = "nchw";
+      ii = nchw.find(keyl);
+      if (ii == std::string::npos)
+        return '\0';
     }
+    dims.push_back(vals[ii]);
+    return kmap[ii];
+  };
+
+  // Determine dimensions.
+  for (size_t i = 0; i < 4; ++i) {
+    filterLayout[i] =
+        convertLayout(filterLayout[i], filterKeys, filterVals, filterDimension);
+    inputLayout[i] =
+        convertLayout(inputLayout[i], inputKeys, inputVals, inputDimension);
+    outputLayout[i] =
+        convertLayout(outputLayout[i], outputKeys, outputVals, outputDimension);
   }
 
   return success();
 }
 
-LogicalResult Conv2dGenerator::genConvModule(
-    std::string &arch, int num_cu, std::string &operation,
-    std::string &inputLayout, std::string &outputLayout,
-    std::string &filterLayout, const SmallVector<int64_t, 4> &filterDimension,
-    const SmallVector<int64_t, 4> &inputDimension,
-    const SmallVector<int64_t, 4> &outputDimension, int dilationHeight,
-    int dilationWidth, int strideHeight, int strideWidth, int paddingHeight,
-    int paddingWidth, ModuleOp &module, OpBuilder &builder,
-    std::string &kernelName, mlir::Type dataType, bool xdlops) {
+LogicalResult Conv2dGenerator::genConvModule(ModuleOp &module,
+                                             OpBuilder &builder) {
+
+  mlir::Type dataType;
+  if (dataTypeStr == "f32" || dataTypeStr == "fp32") {
+    dataType = builder.getF32Type();
+  } else if (dataTypeStr == "f16" || dataTypeStr == "fp16") {
+    dataType = builder.getF16Type();
+  } else if (dataTypeStr == "bf16") {
+    dataType = builder.getIntegerType(16);
+  } else {
+    return failure();
+  }
```

Comment:
Will this be consumed by a `Conv2dOp`? I don't see it getting used somewhere.

Reply:
Just below in constructing the MemRefTypes, it is the element type.

Reply:
Gotcha, thanks!